### PR TITLE
[For discussion] Merge development repo into govuk-puppet

### PR DESCRIPTION
**Pull request by boffbowsh**
As per https://github.gds/gds/development/issues/376, merge
the GOV.UK development repository into govuk-puppet so developers
have a single starting point.

Some of the history of the other repo has been squashed. I found a
point in time where the last of the credentials had been removed
and useful documentation was available. This is also where the repo
most resembles what we use currently.

Steps to create this:

Old repo:
1. New branch at the newest commit we want to lose history for
  `git checkout -b squash-point; git reset --hard 30606ae4486e81d3c2997b511b4028e6e89c98bf`
2. Replace with a new commit containing the current tree
  `git reset $(git commit-tree HEAD^{tree} -m "base commit")`
3. Rebase from first commit we want to keep up to HEAD onto the new commit
  `git rebase -p 7f78f79e0d033abec38f61a2de08524265921914 HEAD --onto squash-point`
4. Remove PR numbers to avoid invalid links
  `git filter-branch --msg-filter 'sed -e "s/Merge pull request #[0-9]*/Merge pull request/g"'`
5. Move all files into `development-vm`
  ```
  git filter-branch -f --index-filter 'git ls-files -s | gsed "s-\t\"*-&development\-vm/-" |
    GIT_INDEX_FILE=$GIT_INDEX_FILE.new \
      git update-index --index-info &&
      mv "$GIT_INDEX_FILE.new" "$GIT_INDEX_FILE"' HEAD
  ```

This repo:

6. Add old repo as an upstream
  `git remote add development-repo /Users/paulbowsher/code/development && git fetch development-repo`
7. Create a new branch
  `git checkout -b merge-development-repo`
8. Merge other repo into this branch and push
  `git merge --allow-unrelated-histories development-repo/squashed && git push origin merge-development-repo`

TODO: 
- [x] Agree approach
- [ ] Merge / rewrite READMEs
- [ ] Update scripts to point at `../../` instead of `../` as needed

### CODE REVIEW

**Comment by surminus** [comment]
I think this makes sense in so far as the development-vm/ directory already existed, and the scripts migrated over are directly related to provisioning using the configuration in the puppet on this repo and that Vagrantfile.

Perhaps in the not so distant future the govuk-puppet code [won't be a requirement to set up a development environment](https://docs.docker.com/compose/), and in that instance a separate repository would more than likely make sense. I think that's worth bearing in mind that this repository could become redundant for a development environment, in which case a fresh "development" repository will be created, but shouldn't stop this change by any means.

**Comment by boffbowsh** [comment]
That's a good point @surminus, I think we can burn that bridge if/when we come to it. It's relatively simple to move it out again later. 
If people are agreed I'll work on the READMEs and scripts.

File: development-vm/README-dev.md [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+should be run in the shell on your Mac, whereas commands that look like this:
+
+    dev$ echo "Linux for human beings"
+
+should be run in the shell on the development VM.
+
+This repository contains tools you'll find useful during development, and scripts for replicating data
+from our integration environment.
+
+The configuration you'll need for your VM lives in a separate repository, govuk-puppet. When we talk about `development-vm`, we mean `govuk_puppet/development-vm`, which is where you'll run vagrant from.
```
Comment (whoojemaflip) [inline development-vm/README-dev.md]: Needs amending

File: development-vm/README-dev.md [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+## 1. Prerequisites and assumptions
+
+  * You are using a Mac
+  * You should have an account on:
+    * GitHub.com (public) and be added to the `alphagov` organisation
+    * GitHub Enterprise (private) and be added to the `gds` organisation
```
Comment (whoojemaflip) [inline development-vm/README-dev.md]: Is this still true?

File: development-vm/README-dev.md [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+be able to ignore some errors.
+
+    dev$ PROC_COUNT=1 ./update-bundler.sh
+
+[bundler]: http://bundler.io/rationale.html
+[pip]: https://pip.pypa.io/en/stable/
+
+## 4. Running the apps
+
+You can run any of the GOV.UK apps from the `/var/govuk/development` directory.
```
Comment (whoojemaflip) [inline development-vm/README-dev.md]: Needs review

File: development-vm/README-dev.md [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+The examples in this section may not work until you've imported production data (see below).
+
+You can use [foreman][foreman] to run a single app. The available apps are defined in the Procfile.
+
+    dev$ cd /var/govuk/development
```
Comment (whoojemaflip) [inline development-vm/README-dev.md]: 👀

File: development-vm/README-dev.md [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+- Licensify
+- Signon
+
+Mapit's database will be downloaded in the Mapit repo and therefore will be not be in the backups folder.
+
+To get production data on to your local vm you will need to either have access
+to integration or a mongo and a mysql export from someone that does. If you
+have integration access then importing the latest data can be done by running
+the following from `development/replication`.
```
Comment (whoojemaflip) [inline development-vm/README-dev.md]: :eyes:

File: development-vm/README-dev.md [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    │   │   └── latest.tbz2
+    │   └── whitehall-mysql-backup-1.backend.integration
+    │       └── latest.tbz2
+    └── postgresql
+        ├── postgresql-master-1.backend.integration
+        |   └── latest.tbz2
+        └── transition-postgresql-master-1.backend.integration
+            └── latest.tbz2
+
+Then, from `development/replication` run.
```
Comment (whoojemaflip) [inline development-vm/README-dev.md]: :eyes:

File: development-vm/README-dev.md [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    -s       Skip downloading the backups (use with -d to load old backups).
+    -r       Reset ignore list. This overrides any default ignores.
+    -i       Databases to ignore. Can be used multiple times, or as a quoted space-delimited list.
+    -n       Don't actually import anything (dry run).
+
+After replicating data a few times, your machine might be running low
+on disk space. This is because the old database dumps aren't cleaned
+up once newer ones have been downloaded. To get over this, it is
+advised to periodically `rm -r` older directories in
+`development/replication/backups`.
```
Comment (whoojemaflip) [inline development-vm/README-dev.md]: :eyes:

File: development-vm/README-dev.md [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+Note: This assumes that you have followed the previous steps, and have your machine's public
+key added to the Puppet repository, and this has been deployed to the environment
+you're trying to access.
+
+While the applications are available directly via the public
+Internet, SSH access to the environment is via a "jumpbox".
+You will need to configure your machine to use this jumpbox. See
+the [instructions in the Ops
+Manual](https://github.gds/pages/gds/opsmanual/2nd-line/technical-setup.html#ssh-config)
```
Comment (whoojemaflip) [inline development-vm/README-dev.md]: :eyes: for future change to this url

File: development-vm/README-dev.md [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+Assuming you have created an SSH keypair and the public key has been distributed,
+and your SSH config is up-to-date, you can connect to machines with:
+
+    $ ssh backend-1.backend.integration
+    $ ssh backend-1.backend.staging
+    $ ssh backend-1.backend.production
+
+## 9. Keeping your VM up to date
+
+There are a few scripts that should be run regularly to keep your VM up to date. In `development` there is `update-git.sh` and `update-bundler.sh` to help keep your projects and their dependencies up
```
Comment (whoojemaflip) [inline development-vm/README-dev.md]: :eyes:

**Comment by tijmenb** [comment]
@boffbowsh @whoojemaflip would it be an idea if we merge this now and do a separate PR for merging the READMEs?

**Comment by boffbowsh** [comment]
I'm ok with that, but it will probably want updating with the latest commits from the development repo first.

**Comment by boffbowsh** [comment]
@whoojemaflip has now re-done the merge with updates from the `development` repo. :+1: because I can't approve my own PR.

**Comment by boffbowsh** [comment]
✨ 🌟 ⭐️ 👍 ⭐️ 🌟 ✨

**Pull request merged**